File: src/lib/nodes/financials.ts
import { ChatGoogleGenerativeAI } from "@langchain/google-genai";
import { AgentState, FinancialAnalysis } from "../types";
import { getFinancialData } from "../tools/alphaVantage";
import { safeParseLlmJson, extractTextContent } from "../utils/parseJson";

export const financialsNode = async (state: AgentState): Promise<Partial<AgentState>> => {
  const companyInfo = state.companyInfo;

  if (!companyInfo?.name) {
    return {
      streamLog: ["❌ Error: No company name available for Financials Node."],
      financialAnalysis: getEmptyAnalysis(),
    };
  }

  const resolvedName = companyInfo.name;
  const ticker = companyInfo.ticker || null;
  const sector = companyInfo.sector || "their sector";

  try {
    // Step 1 — Fetch financial data (graceful if API key missing)
    const financialData = await getFinancialData(ticker);

    // Step 2 — Analyze with Gemini
    const llm = new ChatGoogleGenerativeAI({
      model: "gemini-2.5-flash",
      temperature: 0.1,
      maxOutputTokens: 2048,
    });

    const systemPrompt =
      "You are a CFA-level financial analyst.\nReturn ONLY valid JSON, no markdown, no explanation.";
    const userPrompt = `Analyze this financial data for ${resolvedName}:\n${JSON.stringify(
      financialData,
      null,
      2
    )}\n\nReturn this exact JSON:\n{\n  "revenueGrowthYoY": number | null,\n  "grossMargin": number | null,\n  "netMargin": number | null,\n  "debtToEquity": number | null,\n  "peRatio": number | null,\n  "marketCap": "string | null",\n  "financialHealthScore": number,\n  "financialHealthRationale": "string",\n  "valuationRisk": "low" | "medium" | "high",\n  "valuationRationale": "string",\n  "estimated": boolean\n}\n\nfinancialHealthScore must be 1-10.\nFor private companies with no data, estimate using industry benchmarks for ${sector} and set estimated: true.\nScore conservatively if data is unavailable.`;

    const response = await llm.invoke([
      { role: "system", content: systemPrompt },
      { role: "user", content: userPrompt },
    ]);

    const raw = extractTextContent(response.content);
    const { data: parsedResult, error } = safeParseLlmJson(raw, "FinancialsNode");

    const result = (parsedResult ?? {}) as Record<string, unknown>;
    if (error) {
      console.warn("[FinancialsNode] Using fallback analysis due to parse error.");
    }

    // Safely cast fields
    const toStr = (val: unknown) =>
      val !== null && val !== undefined ? String(val) : null;

    const healthScore =
      typeof result.financialHealthScore === "number" ? result.financialHealthScore : 5;

    const analysis: FinancialAnalysis = {
      revenueGrowthYoY: toStr(result.revenueGrowthYoY),
      grossMargin: toStr(result.grossMargin),
      netMargin: toStr(result.netMargin),
      debtToEquity: result.debtToEquity as number | null,
      peRatio: result.peRatio as number | null,
      marketCap: toStr(result.marketCap),
      financialHealthScore: healthScore,
      financialHealthRationale:
        (result.financialHealthRationale as string) ||
        "No data available to determine health.",
      valuationRisk: (result.valuationRisk as "low" | "medium" | "high") || "medium",
      valuationRationale:
        (result.valuationRationale as string) ||
        "No data available to determine valuation.",
      estimated: !!result.estimated,
    };

    return {
      financialAnalysis: analysis,
      streamLog: [`✅ Financials analyzed. \n  Health score: ${healthScore}/10`],
    };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    return {
      streamLog: [`❌ Error in Financials Node: ${errorMessage}`],
      financialAnalysis: getEmptyAnalysis(),
    };
  }
};

function getEmptyAnalysis(): FinancialAnalysis {
  return {
    revenueGrowthYoY: null,
    grossMargin: null,
    netMargin: null,
    debtToEquity: null,
    peRatio: null,
    marketCap: null,
    financialHealthScore: 5,
    financialHealthRationale: "Failed to fetch or parse financial data.",
    valuationRisk: "medium",
    valuationRationale: "Unknown due to error.",
    estimated: true,
  };
}

File: backend/nodes/financials.py
"""
nodes/financials.py — Financial Analysis Node
Port of src/lib/nodes/financials.ts
"""

import json
import logging
from langchain_groq import ChatGroq
from langchain_core.messages import SystemMessage, HumanMessage

from state import AgentState
from utils.parse_json import safe_parse_llm_json
from tools.alpha_vantage import get_financial_data

logger = logging.getLogger(__name__)


def _empty_analysis() -> dict:
    return {
        "revenueGrowthYoY": None,
        "grossMargin": None,
        "netMargin": None,
        "debtToEquity": None,
        "peRatio": None,
        "marketCap": None,
        "financialHealthScore": 5,
        "financialHealthRationale": "Failed to fetch or parse financial data.",
        "valuationRisk": "medium",
        "valuationRationale": "Unknown due to error.",
        "estimated": True,
    }


def _to_str(val: object) -> object:
    """Return str(val) if not None, else None."""
    return str(val) if val is not None else None


async def financials_node(state: AgentState) -> dict:
    company_info = state.get("companyInfo") or {}

    if not company_info.get("name"):
        return {
            "streamLog": ["⊗ ABORT — No entity resolved for Financials."],
            "financialAnalysis": _empty_analysis(),
        }

    resolved_name = company_info["name"]
    ticker = company_info.get("ticker") or None
    sector = company_info.get("sector") or "their sector"

    try:
        # ── Step 1: Fetch Alpha Vantage data ────────────────────────────────────
        financial_data = await get_financial_data(ticker)

        # ── Step 2: Analyse with Groq ───────────────────────────────────────────
        llm = ChatGroq(
            model="openai/gpt-oss-120b",
            temperature=0.1,
            max_tokens=2048,
        )

        system_prompt = (
            "You are a CFA-level financial analyst.\n"
            "Return ONLY valid JSON, no markdown, no explanation."
        )
        user_prompt = (
            f"Analyze this financial data for {resolved_name}:\n"
            f"{json.dumps(financial_data, indent=2)}\n\n"
            f"Return this exact JSON:\n"
            f'{{\n'
            f'  "revenueGrowthYoY": number | null,\n'
            f'  "grossMargin": number | null,\n'
            f'  "netMargin": number | null,\n'
            f'  "debtToEquity": number | null,\n'
            f'  "peRatio": number | null,\n'
            f'  "marketCap": "string | null",\n'
            f'  "financialHealthScore": number,\n'
            f'  "financialHealthRationale": "string",\n'
            f'  "valuationRisk": "low" | "medium" | "high",\n'
            f'  "valuationRationale": "string",\n'
            f'  "estimated": true or false\n'
            f'}}\n\n'
            f"financialHealthScore must be 1-10.\n"
            f"For private companies with no data, estimate using industry benchmarks for {sector} and set estimated: true.\n"
            f"Score conservatively if data is unavailable."
        )

        response = await llm.ainvoke([
            SystemMessage(content=system_prompt),
            HumanMessage(content=user_prompt),
        ])
        raw = response.content if isinstance(response.content, str) else str(response.content)
        parsed_result, error = safe_parse_llm_json(raw, "FinancialsNode")

        result = parsed_result or {}
        if error:
            logger.warning("[FinancialsNode] Using fallback analysis due to parse error.")

        health_score = result.get("financialHealthScore", 5)
        if not isinstance(health_score, (int, float)):
            health_score = 5

        analysis = {
            "revenueGrowthYoY": _to_str(result.get("revenueGrowthYoY")),
            "grossMargin": _to_str(result.get("grossMargin")),
            "netMargin": _to_str(result.get("netMargin")),
            "debtToEquity": result.get("debtToEquity"),
            "peRatio": result.get("peRatio"),
            "marketCap": _to_str(result.get("marketCap")),
            "financialHealthScore": health_score,
            "financialHealthRationale": result.get("financialHealthRationale") or "No data available to determine health.",
            "valuationRisk": result.get("valuationRisk") or "medium",
            "valuationRationale": result.get("valuationRationale") or "No data available to determine valuation.",
            "estimated": bool(result.get("estimated", False)),
        }

        return {
            "financialAnalysis": analysis,
            "streamLog": [f"◈ FINANCIALS PROCESSED. \n  Health score: {health_score}/10"],
        }

    except Exception as exc:
        error_message = str(exc)
        return {
            "streamLog": [f"⊗ FINANCIALS FAULT — {error_message}"],
            "financialAnalysis": _empty_analysis(),
        }
